bootstrap/src/hmc/scan.ml

```diff
@@ -0,0 +1,4334 @@
+      | _ -> fn nesting cursor t
+    ) nesting cursor t
+  end in
+  fn 1 cursor t
+
+let operator_cps = "-+*/%@!$<=>|:.~?"
+let operator_set = set_of_cps operator_cps
+
+let operator_map = (
+  let open AbstractToken in
+  Map.of_alist (module String) [
+    ("|", Tok_bar);
+    ("->", Tok_arrow);
+    (">", Tok_gt);
+    (">->", Tok_parrow);
+  ])
+
+let operator fop _pcursor cursor t =
+  let accept_operator fop cursor t = begin
+    let op = str_of_cursor cursor t in
+    match Map.get op operator_map with
+    | Some tok -> accept tok cursor t
+    | None -> accept (fop op) cursor t
+  end in
+  let rec fn fop cursor t = begin
+    match Text.Cursor.next_opt cursor with
+    | None -> accept_operator fop cursor t
+    | Some (cp, cursor') -> begin
+        match Set.mem cp operator_set with
+        | true -> fn fop cursor' t
+        | false -> accept_operator fop cursor t
+      end
+  end in
+  fn fop cursor t
+
+let ident_set = set_of_cps
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_'"
+
+let ident ~f_accept cursor t =
+  let rec fn cursor t = begin
+    match Text.Cursor.next_opt cursor with
+    | None -> f_accept cursor t
+    | Some (cp, cursor') -> begin
+        match Set.mem cp ident_set with
+        | true -> fn cursor' t
+        | false -> f_accept cursor t
+      end
+  end in
+  fn cursor t
+
+let uident _pcursor cursor t =
+  ident ~f_accept:(fun cursor t ->
+    let uident_str = str_of_cursor cursor t in
+    accept (AbstractToken.of_uident_str uident_str) cursor t
+  ) cursor t
+
+let cident _pcursor cursor t =
+  ident ~f_accept:(fun cursor t ->
+    let cident_str = str_of_cursor cursor t in
+    accept (Tok_cident cident_str) cursor t
+  ) cursor t
+
+let accum_cp_of_nat ~accum_cp ~accum_mal nat accum base past t =
+  let nat_to_cp_opt nat = begin
+    match Nat.to_uns_opt nat with
+    | None -> None
+    | Some u -> Codepoint.of_uns_opt u
+  end in
+  match nat_to_cp_opt nat with
+  | None -> accum_mal (invalid_unicode base past t) accum
+  | Some cp -> accum_cp cp accum
+
+module Codepoint_ : sig
+  type umap =
+    | UMapUscore
+    | UMapDigit
+    | UMapRcurly
+    | UMapDitto
+    | UMapTick
+
+  val u_map: (Codepoint.t, umap, Codepoint.cmper_witness) Map.t
+  val codepoint: Text.Cursor.t -> Text.Cursor.t -> t -> t * ConcreteToken.t
+end = struct
+  type umap =
+    | UMapUscore
+    | UMapDigit
+    | UMapRcurly
+    | UMapDitto
+    | UMapTick
+
+  let u_map = map_of_cps_alist [
+    ("_", UMapUscore);
+    ("0123456789abcdef", UMapDigit);
+    ("}", UMapRcurly);
+    ("\"", UMapDitto);
+    ("'", UMapTick);
+  ]
+
+  type bmap =
+    | BMapU
+    | BMapT
+    | BMapN
+    | BMapR
+    | BMapTick
+    | BMapBslash
+    | BMapNewline
+
+  let bslash_map = map_of_cps_alist [
+    ("u", BMapU);
+    ("t", BMapT);
+    ("n", BMapN);
+    ("r", BMapR);
+    ("'", BMapTick);
+    ("\\", BMapBslash);
+    ("\n", BMapNewline);
+  ]
```

Comment:
Although this map is an optimization for the OCaml version of the scanner, Hemlock will be able to generate efficient pattern matching trees when all the cases are exposed. In fact, for the Hemlock version, all the maps (including ones like `u_map` above) may be better written as plain pattern matches.